### PR TITLE
PERF-3726 external phase coordinator.

### PR DESCRIPTION
**Pull request by jimoleary**
**What:**
The genny portion of the phase event handling. When run with './run-genny' directly, the named pipes are not created so genny does not emit any events. When run under DSI, the named pipes are available so genny writes phase to the out named pipe and waits for an ack from dsi before continuing.

The code addes start and end phase events. These could be trimmed to just a single event type.

The phase number was added as a parameter to the callback because there is already a write lock taken out, so it is not possible to read the phase number through the accessor. Adding the number to the start and stop seems reasonable to me.

**Patches:** 
[updated sys perf patch](https://spruce.mongodb.com/version/64049adb5623436fe78fec71/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)
~~[sys perf](https://spruce.mongodb.com/version/6402097e0ae6061e081c36be/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC).~~

**Other PRs:**
[DSI](https://github.com/10gen/dsi/pull/1227)

### CODE REVIEW

File: src/gennylib/src/Orchestrator.cpp [hunk 1]
```diff
@@ -136,8 +136,12 @@ bool Orchestrator::awaitPhaseEnd(bool block, int removeTokens) {
     // compare with >= rather than ==.
 
     if (_currentTokens <= 0) {
+        const auto currentPhase = _current;
         ++_current;
-        BOOST_LOG_TRIVIAL(info) << "Ended phase " << (this->_current - 1);
+        BOOST_LOG_TRIVIAL(info) << "Ended phase " << currentPhase;
```
Comment (dalyd) [inline src/gennylib/src/Orchestrator.cpp:141]: This log message is going to change by 1. Is that intentional?

File: src/gennylib/src/Orchestrator.cpp [hunk 2]
```diff
@@ -136,8 +136,12 @@ bool Orchestrator::awaitPhaseEnd(bool block, int removeTokens) {
...
+        for (auto&& cb : _postPhaseHooks) {
+            cb(this, currentPhase);
```
Comment (dalyd) [inline src/gennylib/src/Orchestrator.cpp:143]: Do you want the next phase number for the call back, or the current? could you comment that somewhere?

File: src/gennylib/src/context.cpp [hunk 3]
```diff
@@ -205,4 +214,55 @@ bool PhaseContext::isNop() const {
...
+}
+void ExternalPhaseCoordinator::onPhaseStart(PhaseNumber phase){
+    std::stringstream event;
+    event << R"({ "message": ")" << "Beginning phase" << R"(","phase":)"
+          <<  phase << R"(,"request":)" <<  1 << "}" << std::endl;
+    _onPhase(event.str());
+}
+void ExternalPhaseCoordinator::onPhaseStop(PhaseNumber phase){
+    std::stringstream event;
+    event << R"({ "message": ")" << "Ended phase" << R"(","phase":)"
```
Comment (dalyd) [inline src/gennylib/src/context.cpp:263]: I think you are going to be off 1 on the ended phase based on the above.

File: src/gennylib/src/context.cpp [hunk 4]
```diff
@@ -205,4 +214,55 @@ bool PhaseContext::isNop() const {
...
+    if(connect(sd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+        BOOST_LOG_TRIVIAL(warning) << "connect('" << ipaddress << "',"
+                                   << port << "') failed --> '"<< ::strerror(errno) << "'";
+        return -1;
+    }
+    return sd;
+}
+
+ExternalPhaseCoordinator::ExternalPhaseCoordinator(std::string host, int port) :
+      m_host(host), m_port(port), m_socket(connect(host, port)){
```
Comment (dalyd) [inline src/gennylib/src/context.cpp:237]: Not to self, this is where the connect function is called and the ip connection is made.

File: src/gennylib/src/Orchestrator.cpp (hunk 2 above)
Comment (dalyd) [reply src/gennylib/src/Orchestrator.cpp:143]: Looking below I think you want to call this line before incrementing current on line 140.

**Review by dalyd — CHANGES_REQUESTED** [review]

**Review by jimoleary — COMMENTED** [review]
I think the confusion here is due to the vector names. I'll change  them to be more clear and adjust the code to reinforce this (plus add some comments).

File: src/gennylib/src/context.cpp (hunk 4 above)
Comment (dalyd) [reply src/gennylib/src/context.cpp:239]: A comment would be good. It's clear, but I also read right past it and then went searching for it.

Comment (jimoleary) [reply src/gennylib/src/context.cpp:241]: done.

File: src/gennylib/src/Orchestrator.cpp (hunk 2 above)
Comment (dalyd) [reply src/gennylib/src/Orchestrator.cpp]: All good now. Thanks. I still don't think the old code works, but that's no longer relevant :)

**Review by dalyd — APPROVED** [review]
Thanks for the updates. Much clearer to me now.

**Comment by jimoleary** [comment]
evergreen merge

**Pull request merged**